perf_test.go

```diff
@@ -46,28 +47,118 @@ func TestPerfReader(t *testing.T) {
 }
 
 func TestPerfReaderLostSample(t *testing.T) {
-	coll, err := LoadCollection("testdata/perf_output.elf")
+	// To generate a lost sample perf record:
+	//
+	// 1. Fill the perf ring buffer almost completely, with the output_large program.
+	//    The buffer is sized in number of pages, which are architecture dependant.
+	//
+	// 2. Write an extra event that doesn't fit in the space remaining.
+	//
+	// 3. Write a smaller event that does fit, with output_single program.
+	//    Lost sample records are generated opportunistically, when the kernel
+	//    is writing an event and realizes that there were events lost previously.
+	//
+	// The event size is hardcoded in the test BPF programs, there's no way
+	// to parametrize it without rebuilding the programs.
+	//
+	// The event size needs to be selected so that, for any page size, there are at least
+	// 48 bytes left in the perf ring page after filling it with a whole number of events:
+	//
+	//  - PERF_RECORD_LOST: 8 (perf_event_header) + 16 (PERF_RECORD_LOST)
+	//
+	//  - output_single: 8 (perf_event_header) + 4 (size) + 5 (payload) + 7 (padding to 64bits)
+	//
+	// By selecting an event size of the form 2^n + 2^(n+1), for any page size 2^(n+m), m >= 0,
+	// the number of bytes left, x, after filling a page with a whole number of events is:
+	//
+	//                     2^(n+m)                            2^n * 2^m
+	//  x = 2^n * frac(---------------) <=> x = 2^n * frac(---------------)
+	//                  2^n + 2^(n+1)                       2^n + 2^n * 2
+	//
+	//                                                        2^n * 2^m
+	//                                  <=> x = 2^n * frac(---------------)
+	//                                                      2^n * (1 + 2)
+	//
+	//                                                      2^m
+	//                                  <=> x = 2^n * frac(-----)
+	//                                                       3
+	//
+	//                                                1                2
+	//                                  <=> x = 2^n * -  or  x = 2^n * -
+	//                                                3                3
+	//
+	// Selecting n = 6, we have:
+	//
+	//  x = 64  or  x = 128, no matter the page size 2^(6+m)
+	//
+	//  event size = 2^6 + 2^7 = 192
+	//
+	// Accounting for perf headers, output_large uses a 180 byte payload:
+	//
+	//  8 (perf_event_header) + 4 (size) + 180 (payload)
+	const eventSize = 192
+
+	pageSize := os.Getpagesize()
+
+	remainder := pageSize % eventSize
+	if !(remainder == 64 || remainder == 128) {
+		// Page size isn't 2^(6+m), m >= 0
+		t.Fatal("unsupported page size:", pageSize)
+	}
+
+	spec, err := LoadCollectionSpec("testdata/perf_output.elf")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Single perf map for output_large and output_single
+	perfMap, err := NewMap(&MapSpec{
```

Comment:
Couldn't you just DetachMap after NewCollection?

Reply:
Duh, I completely zapped that one ELF, with two entrypoints == one map.. Fixed!